deploy/aws/java11Exec/src/main/java/poseidon/App.java

```diff
@@ -0,0 +1,136 @@
+    private boolean disableOutput = false;
+
+    public APIGatewayProxyResponseEvent handleRequest(final APIGatewayV2WebSocketEvent input, final Context context) {
+        APIGatewayV2WebSocketEvent.RequestContext ctx = input.getRequestContext();
+        this.gwClient = AmazonApiGatewayManagementApiClientBuilder.standard()
+                .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration("https://" + ctx.getDomainName() + "/" + ctx.getStage(), Regions.EU_CENTRAL_1.getName()))
+                .build();
+        this.connectionID = ctx.getConnectionId();
+        this.disableOutput = input.getHeaders() != null && input.getHeaders().containsKey(disableOutputHeaderKey) && Boolean.parseBoolean(input.getHeaders().get(disableOutputHeaderKey));
+        AwsFunctionRequest execution = gson.fromJson(input.getBody(), AwsFunctionRequest.class);
+
+        try {
+            File workingDirectory = this.writeFS(execution.files);
+
+            ProcessBuilder pb = new ProcessBuilder(execution.cmd).redirectErrorStream(true);
+            pb.directory(workingDirectory);
+            Process p = pb.start();
+            InputStream stdout = p.getInputStream(), stderr = p.getErrorStream();
+            this.forwardOutput(p, stdout, stderr);
+            p.destroy();
+            return new APIGatewayProxyResponseEvent().withStatusCode(200);
+        } catch (Exception e) {
+            this.sendMessage(e.toString().getBytes(StandardCharsets.UTF_8), WebSocketMessageType.WebSocketOutputError, 0);
+            return new APIGatewayProxyResponseEvent().withBody(e.toString()).withStatusCode(500);
+        }
+    }
+
+    // writeFS writes the files to the local filesystem.
+    private File writeFS(Map<String, String> files) throws IOException {
+        File workspace = Files.createTempDirectory("workspace").toFile();
+        for (Map.Entry<String, String> entry : files.entrySet()) {
+            File f = new File(workspace, entry.getKey());
+
+            boolean ok = f.getParentFile().mkdirs();
+            if (!ok) throw new IOException("Cannot create parent directories.");
+
+            ok = f.createNewFile();
+            if (!ok) throw new IOException("Cannot create file.");
+
+            Files.write(f.toPath(), Base64.getDecoder().decode(entry.getValue()));
+        }
+        return workspace;
+    }
+
+    // forwardOutput sends the output of the process to the WebSocket connection.
+    private void forwardOutput(Process p, InputStream stdout, InputStream stderr) throws InterruptedException, IOException {
+        while (p.isAlive() || stdout.available()>0 || stderr.available()>0) {
+            while(stdout.available()>0) this.sendMessage(stdout.readAllBytes(), WebSocketMessageType.WebSocketOutputStdout, 0);
+            while(stderr.available()>0) this.sendMessage(stderr.readAllBytes(), WebSocketMessageType.WebSocketOutputStderr, 0);
+            Thread.sleep(50);
```

Comment:
Hm, why are we putting the thread to sleep here? Is that due to prevent busy waiting or some race condition fighting?

Reply:
This should prevent busy waiting. Now we use the Scanner class to so.
(Subclassing the InputStream has no effect since the Process create it's own implementation of the InputStream.)